skll/data/readers.py

```diff
@@ -765,68 +801,54 @@ def _sub_read(self, f):
-                row_type = split_header[0].lower()
-                if row_type == '@attribute':
-                    # Add field name to list
-                    field_name = split_header[1]
-                    field_names.append(field_name)
-                    # Check if we're doing regression
-                    if field_name == self.label_col:
-                        self.regression = (len(split_header) > 2 and
-                                           split_header[2] == 'numeric')
-                # Save relation if specified
-                elif row_type == '@relation':
-                    self.relation = split_header[1]
-                # Stop at data
-                elif row_type == '@data':
-                    break
-                # Skip other types of rows (relations)
-
-        # Create header for CSV
-        if PY2:
-            io_type = BytesIO
-        else:
-            io_type = StringIO
-        with io_type() as field_buffer:
-            csv.writer(field_buffer, dialect='arff').writerow(field_names)
-            field_str = field_buffer.getvalue()
+        with open(f, 'r' if PY3 else 'rb') as buff:
+
+            lines = [UnicodeDammit(line.strip(), ['utf-8', 'windows-1252']).unicode_markup
+                     if not isinstance(line, text_type) and PY2
+                     else line.strip()
+                     for line in buff if line.strip()]
+
+        # find the row index starting with data; the line below this
+        # is where the data should actually begin
+        data_idx = lines.index('@data')
+
+        df = self.split_with_quotes('\n'.join(lines[data_idx + 1:]), delimiter=',')
+
+        # get the column names from the attribute
+        # rows, and add them to the columns list
+        columns = []
+        for row in lines[:data_idx]:
+
+            row_series = self.split_with_quotes(row)
+            if row_series.loc[0, 0] == '@attribute':
+                column = row_series.loc[0, 1]
+                columns.append(column)
+
+                # if the column is the label column,
+                # and the type is 'numeric', set regression
+                # to True; otherwise False
+                if column == self.label_col:
+                    self.regression = row_series.loc[0, 2] == 'numeric'
 
-        # Set label_col to be the name of the last field, since that's standard
-        # for ARFF files
-        if self.label_col != field_names[-1]:
-            self.label_col = None
+            # if the relation attribute exists, then
+            # add it to the relation instance variable
+            elif row_series.loc[0, 0] == '@relation':
+                self.relation = row_series.loc[0, 1]
 
-        # Process data as CSV file
-        return super(ARFFReader, self)._sub_read(chain([field_str], f))
+        df.columns = columns
+        return self._parse_dataframe(df, self.id_col, self.label_col)
 
 
-class TSVReader(DelimitedReader):
+class TSVReader(Reader):
```

Comment:
It might be nice if we could make `TSVReader` a sub-class of `CSVReader` with an instance attribute like `sep` set to `"\t"` (the default, defined in `CSVReader`, would be `","`). Then, `sep=self.sep` could replace e.g. `sep='\t'` in `pd.read_csv`. It would reduce the duplication of the two classes by a lot.

Reply:
The same thing could be done for `TSVWriter`/`CSVWriter`.